File: app/src/main/java/com/example/boulderjournal/AddRouteActivity.java
package com.example.boulderjournal;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.FileProvider;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.example.boulderjournal.Utils.Utilities;
import com.example.boulderjournal.data.AppDatabase;
import com.example.boulderjournal.data.RouteEntry;
import com.squareup.picasso.Picasso;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;


public class AddRouteActivity extends AppCompatActivity {

    public static final String EXTRA_ROUTE_ID = "extraTaskId";
    public static final String INSTANCE_ROUTE_ID = "instanceTaskId";
    private static final int DEFAULT_ROUTE_ID = -1;


    private static final String DATE_FORMAT = "dd/MM/yyy";
    private SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_FORMAT, Locale.getDefault());


    private EditText mRouteName;
    private EditText mRoom;
    private EditText mWall;
    private EditText mNotes;

    private RadioGroup mRouteColourGroup;
    private RadioButton mRouteColourButtom;
    private View mColorSwatch;

    private TextView mDateTV;
    private TextView mRouteNameTV;
    private TextView mRouteColourTV;
    private TextView mRoomTV;
    private TextView mWallTV;
    private TextView mNotesTV;
    private ImageView wallPhotoImageView;

    private String routeName;
    private String routeColour;
    private String room;
    private String wall;
    private String notes;
    private Date date;
    private Boolean completed = false;

    private boolean editMenuCheck = false;
    private boolean readyUpdateCheck = false;

    private boolean readyAddDb = false;

    private MenuItem editMenuItem;
    private MenuItem addMenuItem;

    private RouteEntry route;


    private Button mPhotoIntentButton;
    static final int REQUEST_IMAGE_CAPTURE = 1;
    private String currentPhotoPath;
    private Uri climbWallUri;

    private int mRouteId = DEFAULT_ROUTE_ID;

    private AppDatabase mDb;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTheme(R.style.AppTheme);
        setContentView(R.layout.activity_add_route);
        initViews();
        mDb = AppDatabase.getInstance(getApplicationContext());

        if (savedInstanceState != null && savedInstanceState.containsKey(INSTANCE_ROUTE_ID)) {
            mRouteId = savedInstanceState.getInt(INSTANCE_ROUTE_ID, DEFAULT_ROUTE_ID);

        }



        Intent intent =  getIntent();
        if(intent != null & intent.hasExtra(EXTRA_ROUTE_ID)){
            editMenuCheck = true;
            initStaticViews();

            final int routeID = intent.getIntExtra(EXTRA_ROUTE_ID, DEFAULT_ROUTE_ID) ;

            AppExecutors.getInstance().diskIO().execute(new Runnable() {
                @Override
                public void run() {
                     route = mDb.routeDao().loadRouteById(routeID);
                    populateStaticUI(route);
                }
            });

        }

        wallPhotoImageView = (ImageView)findViewById(R.id.capturedPhoto);
        mPhotoIntentButton = (Button)findViewById(R.id.updateButton);
        mPhotoIntentButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dispatchTakePictureIntent();

            }
        });


    }

    public void setValues(){
        routeName = mRouteName.getText().toString();
        int selectedId = mRouteColourGroup.getCheckedRadioButtonId();
        mRouteColourButtom = (RadioButton)findViewById(selectedId);
        if(mRouteColourButtom != null) {
            routeColour = mRouteColourButtom.getText().toString();
        }
        room = mRoom.getText().toString();
        wall = mWall.getText().toString();
        notes = mNotes.getText().toString();
        String completedStr = completed.toString();
        if(date == null){  date = new Date();}


        if (routeName.length() == 0 || routeColour.length() == 0 || room.length() == 0 || wall.length() == 0 || notes.length() == 0  ) {
            Toast.makeText(getBaseContext(), "Please fill in all fields", Toast.LENGTH_LONG).show();
            return;
        }else if(route == null){
           route = new RouteEntry(routeName, routeColour, room, wall, notes, completedStr, date);
           readyAddDb = true;
        }

       else{
           route.setRouteName(routeName);
           route.setRouteColour(routeColour);
           route.setRoom(room);
           route.setWall(wall);
           route.setNote(notes);
           readyAddDb = true;
       }
    }

     public void onAddRoute(){

         setValues();
         if(readyAddDb) {
             AppExecutors.getInstance().diskIO().execute(new Runnable() {
                 @Override
                 public void run() {
                     mDb.routeDao().insertRoute(route);
                     finish();
                 }
             });
         }
     }

     public void onUpdateRoute(){
        setValues();
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                mDb.routeDao().updateRoute(route);
                finish();
            }
        });

     }




    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        editMenuItem = menu.findItem(R.id.edit_menu_item);
        addMenuItem = menu.findItem(R.id.add);

        if(editMenuCheck){
        editMenuItem.setVisible(editMenuCheck);
        addMenuItem.setVisible(false);
        }
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.add_route_menu_button, menu);



        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.add) {
            onAddRoute();
            return true;
        }else if(id == R.id.edit_menu_item && editMenuCheck){
            populateEditableUI(route);
        }else if(id == R.id.edit_menu_item && readyUpdateCheck){
            Toast.makeText(getBaseContext(), "Ready", Toast.LENGTH_LONG).show();
            onUpdateRoute();
        }
        return super.onOptionsItemSelected(item);
    }

    public void initViews(){

        String newFormattedDate = dateFormat.format(new Date());
        mDateTV  = (TextView)findViewById(R.id.date_added);
        mDateTV.setText(newFormattedDate);

        mRouteName = (EditText) findViewById(R.id.input_route_name);
        mRouteColourGroup = (RadioGroup)findViewById(R.id.route_colour_group);
        mRoom = (EditText)findViewById(R.id.input_room);
        mWall = (EditText)findViewById(R.id.input_wall);
        mNotes = (EditText)findViewById(R.id.route_note);
    }

    public void initStaticViews(){

        mRouteNameTV = (TextView)findViewById(R.id.view_route_name);
        mRouteColourTV = (TextView)findViewById(R.id.view_route_colour);
        mColorSwatch = (View) findViewById(R.id.color_swatch);
        mRoomTV = (TextView)findViewById(R.id.view_room);
        mWallTV = (TextView)findViewById(R.id.view_wall);
        mNotesTV = (TextView)findViewById(R.id.view_route_notes);



    }

    public void populateStaticUI(RouteEntry routeEntry){

        String formattedDate = dateFormat.format(routeEntry.getUpdatedAt());
        mDateTV.setText(formattedDate);

        mRouteNameTV.setText(routeEntry.getRouteName());
        mRouteNameTV.setVisibility(View.VISIBLE);
        mRouteName.setVisibility(View.GONE);

        String setRouteColor = routeEntry.getRouteColour();
        mRouteColourTV.setVisibility(View.VISIBLE);
        mRouteColourTV.setText(setRouteColor);
        mRouteColourGroup.setVisibility(View.GONE);
        mColorSwatch.setVisibility(View.VISIBLE);
        int getRouteColourInt = Utilities.getColor(setRouteColor, this);
        mColorSwatch.setBackgroundColor(getRouteColourInt);



        mRoomTV.setText(routeEntry.getRoom());
        mRoomTV.setVisibility(View.VISIBLE);
        mRoom.setVisibility(View.GONE);


        mWallTV.setText(routeEntry.getWall());
        mWallTV.setVisibility(View.VISIBLE);
        mWall.setVisibility(View.GONE);


        mNotesTV.setText(routeEntry.getNote());
        mNotesTV.setVisibility(View.VISIBLE);
        mNotes.setVisibility(View.GONE);


    }

    public void populateEditableUI(RouteEntry routeEntry){

        mRouteName.setText(routeEntry.getRouteName());
        mRouteName.setVisibility(View.VISIBLE);
        mRouteNameTV.setVisibility(View.GONE);

        mRouteColourTV.setVisibility(View.GONE);
        mRouteColourGroup.setVisibility(View.VISIBLE);
        getRadioBox(routeEntry.getRouteColour());
        mColorSwatch.setVisibility(View.GONE);


        mRoom.setText(routeEntry.getRoom());
        mRoom.setVisibility(View.VISIBLE);
        mRoomTV.setVisibility(View.GONE);


        mWall.setText(routeEntry.getWall());
        mWall.setVisibility(View.VISIBLE);
        mWallTV.setVisibility(View.GONE);


        mNotes.setText(routeEntry.getNote());
        mNotes.setVisibility(View.VISIBLE);
        mNotesTV.setVisibility(View.GONE);

        editMenuCheck = false;
        readyUpdateCheck = true;
        editMenuItem.setTitle(R.string.update);
    }

    public void getRadioBox(String color){
        String newColourStr = color.toLowerCase();
        switch (newColourStr) {
            case "blue":
                ((RadioGroup) findViewById(R.id.route_colour_group)).check(R.id.blueButton);
                break;
            case "pink":
                ((RadioGroup) findViewById(R.id.route_colour_group)).check(R.id.pinkButton);
                break;
            case "orange":
                ((RadioGroup) findViewById(R.id.route_colour_group)).check(R.id.orangeButton);
                break;
            case "yellow":
                ((RadioGroup) findViewById(R.id.route_colour_group)).check(R.id.yellowButton);
        }
    }





    private void dispatchTakePictureIntent() {
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        // Ensure that there's a camera activity to handle the intent
        if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
            // Create the File where the photo should go
            File photoFile = null;
            try {
                photoFile = createImageFile();
            } catch (IOException ex) {
              return;
            }
            if (photoFile != null) {
                Toast.makeText(this,"haha", Toast.LENGTH_SHORT);
                Uri photoURI = FileProvider.getUriForFile(Objects.requireNonNull(getApplicationContext()),
                        BuildConfig.APPLICATION_ID + ".fileprovider", photoFile);
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
                startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
                climbWallUri = photoURI;



            }
        }
    }




    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == RESULT_OK) {
            if(climbWallUri != null) {
                Picasso.get().load(climbWallUri).into(wallPhotoImageView);
            }else{
                String url = "https://p2.piqsels.com/preview/385/249/229/rock-climber-rope-mounatin.jpg";
                Picasso.get().load(url).into(wallPhotoImageView);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);

    }

    private File createImageFile() throws IOException {
        // Create an image file name
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
        String imageFileName = "JPEG_" + timeStamp + "_";
        File storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File image = File.createTempFile(
                imageFileName,  /* prefix */
                ".jpg",         /* suffix */
                storageDir      /* directory */
        );

        // Save a file: path for use with ACTION_VIEW intents
        currentPhotoPath = image.getAbsolutePath();
        return image;
    }
}

File: app/src/main/java/com/example/boulderjournal/addRoute/AddRouteActivity.kt
package com.example.boulderjournal.addRoute

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import com.example.boulderjournal.AppExecutors.Companion.instance
import com.example.boulderjournal.BuildConfig
import com.example.boulderjournal.R
import com.example.boulderjournal.Utils.Utilities
import com.example.boulderjournal.data.AppDatabase
import com.example.boulderjournal.data.AppDatabase.Companion.getInstance
import com.example.boulderjournal.data.RouteEntry
import com.squareup.picasso.Picasso
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

class AddRouteActivity : AppCompatActivity() {
    private val dateFormat = SimpleDateFormat(DATE_FORMAT, Locale.getDefault())
    private var mRouteName: EditText? = null
    private var mRoom: EditText? = null
    private var mWall: EditText? = null
    private var mNotes: EditText? = null
    private var mRouteColourGroup: RadioGroup? = null
    private var mRouteColourButtom: RadioButton? = null
    private var mColorSwatch: View? = null
    private var mDateTV: TextView? = null
    private var mRouteNameTV: TextView? = null
    private var mRouteColourTV: TextView? = null
    private var mRoomTV: TextView? = null
    private var mWallTV: TextView? = null
    private var mNotesTV: TextView? = null
    private var wallPhotoImageView: ImageView? = null
    private var routeName: String? = null
    private var routeColour: String? = null
    private var room: String? = null
    private var wall: String? = null
    private var notes: String? = null
    private var imageURIstr: String? = null
    private var date: Date? = null
    private val completed = false
    private var editMenuCheck = false
    private var readyUpdateCheck = false
    private var readyAddDb = false
    private var editMenuItem: MenuItem? = null
    private var addMenuItem: MenuItem? = null
    private var route: RouteEntry? = null
    private var mPhotoIntentButton: Button? = null
    private var currentPhotoPath: String? = null
    private var climbWallUri: Uri? = null
    private var mRouteId = DEFAULT_ROUTE_ID
    private var mDb: AppDatabase? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme(R.style.AppTheme)
        setContentView(R.layout.activity_add_route)
        initViews()
        mDb = getInstance(applicationContext)
        if (savedInstanceState != null && savedInstanceState.containsKey(INSTANCE_ROUTE_ID)) {
            mRouteId = savedInstanceState.getInt(INSTANCE_ROUTE_ID, DEFAULT_ROUTE_ID)
        }
        val intent = intent
        if (intent.hasExtra(EXTRA_ROUTE_ID) and (intent != null)) {
            editMenuCheck = true
            initStaticViews()
            val routeID = intent.getIntExtra(EXTRA_ROUTE_ID, DEFAULT_ROUTE_ID)
            instance!!.diskIO().execute {
                route = mDb!!.routeDao().loadRouteById(routeID)
                populateStaticUI(route)
            }
            delaySetImage()
        }
        mPhotoIntentButton!!.setOnClickListener { dispatchTakePictureIntent() }
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    fun setValues() {
        routeName = mRouteName!!.text.toString()
        val selectedId = mRouteColourGroup!!.checkedRadioButtonId
        mRouteColourButtom = findViewById<View>(selectedId) as RadioButton
        if (mRouteColourButtom != null) {
            routeColour = mRouteColourButtom!!.text.toString()
        }
        room = mRoom!!.text.toString()
        wall = mWall!!.text.toString()
        notes = mNotes!!.text.toString()
        val completedStr = completed.toString()
        if (climbWallUri != null) {
            imageURIstr = climbWallUri.toString()
        }
        if (date == null) {
            date = Date()
        }
        if (routeName!!.length == 0 || routeColour!!.length == 0 || room!!.length == 0 || wall!!.length == 0 || notes!!.length == 0) {
            Toast.makeText(baseContext, "Please fill in all fields", Toast.LENGTH_LONG).show()
            return
        } else if (route == null) {
            route = RouteEntry(0, routeName, routeColour, room, wall, notes, imageURIstr, completedStr, date)
            readyAddDb = true
        } else {
            route!!.routeName = routeName
            route!!.routeColour = routeColour
            route!!.room = room
            route!!.wall = wall
            route!!.note = notes
            route!!.setmImageLocation(imageURIstr)
            readyAddDb = true
        }
    }

    fun onAddRoute() {
        setValues()
        if (readyAddDb) {
            instance!!.diskIO().execute {
                mDb!!.routeDao().insertRoute(route!!)
                finish()
            }
        }
    }

    fun onUpdateRoute() {
        setValues()
        instance!!.diskIO().execute { mDb!!.routeDao().updateRoute(route!!) }
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        editMenuItem = menu.findItem(R.id.edit_menu_item)
        addMenuItem = menu.findItem(R.id.add)
        if (editMenuCheck) {
            editMenuItem?.setVisible(editMenuCheck)
            addMenuItem?.setVisible(false)
        }
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.add_route_menu_button, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.add) {
            onAddRoute()
            return true
        } else if (id == R.id.edit_menu_item && editMenuCheck) {
            populateEditableUI(route)
        } else if (id == R.id.edit_menu_item && readyUpdateCheck) {
            onUpdateRoute()
            refactorUIonUpdateRoute()
        }
        return super.onOptionsItemSelected(item)
    }

    fun initViews() {
        val newFormattedDate = dateFormat.format(Date())
        mDateTV = findViewById<View>(R.id.date_added) as TextView
        mDateTV!!.text = newFormattedDate
        mRouteName = findViewById<View>(R.id.input_route_name) as EditText
        mRouteColourGroup = findViewById<View>(R.id.route_colour_group) as RadioGroup
        mRoom = findViewById<View>(R.id.input_room) as EditText
        mWall = findViewById<View>(R.id.input_wall) as EditText
        mNotes = findViewById<View>(R.id.route_note) as EditText
        wallPhotoImageView = findViewById<View>(R.id.capturedPhoto) as ImageView
        mPhotoIntentButton = findViewById<View>(R.id.updateButton) as Button
    }

    fun initStaticViews() {
        mRouteNameTV = findViewById<View>(R.id.view_route_name) as TextView
        mRouteColourTV = findViewById<View>(R.id.view_route_colour) as TextView
        mColorSwatch = findViewById(R.id.color_swatch) as View
        mRoomTV = findViewById<View>(R.id.view_room) as TextView
        mWallTV = findViewById<View>(R.id.view_wall) as TextView
        mNotesTV = findViewById<View>(R.id.view_route_notes) as TextView
    }

    fun populateStaticUI(routeEntry: RouteEntry?) {
        val formattedDate = dateFormat.format(routeEntry!!.updatedAt)
        mDateTV!!.text = formattedDate
        mRouteNameTV!!.text = routeEntry.routeName
        mRouteNameTV!!.visibility = View.VISIBLE
        mRouteName!!.visibility = View.GONE
        val setRouteColor = routeEntry.routeColour
        mRouteColourTV!!.visibility = View.VISIBLE
        mRouteColourTV!!.text = setRouteColor
        mRouteColourGroup!!.visibility = View.GONE
        mColorSwatch!!.visibility = View.VISIBLE
        val getRouteColourInt = Utilities.getColor(setRouteColor, this)
        mColorSwatch!!.setBackgroundColor(getRouteColourInt)
        mRoomTV!!.text = routeEntry.room
        mRoomTV!!.visibility = View.VISIBLE
        mRoom!!.visibility = View.GONE
        mWallTV!!.text = routeEntry.wall
        mWallTV!!.visibility = View.VISIBLE
        mWall!!.visibility = View.GONE
        mNotesTV!!.text = routeEntry.note
        mNotesTV!!.visibility = View.VISIBLE
        mNotes!!.visibility = View.GONE
        mPhotoIntentButton!!.visibility = View.GONE
        setSavedImageIfPresent()
    }

    fun populateEditableUI(routeEntry: RouteEntry?) {
        mRouteName!!.setText(routeEntry!!.routeName)
        mRouteName!!.visibility = View.VISIBLE
        mRouteNameTV!!.visibility = View.GONE
        mRouteColourTV!!.visibility = View.GONE
        mRouteColourGroup!!.visibility = View.VISIBLE
        getRadioBox(routeEntry.routeColour)
        mColorSwatch!!.visibility = View.GONE
        mRoom!!.setText(routeEntry.room)
        mRoom!!.visibility = View.VISIBLE
        mRoomTV!!.visibility = View.GONE
        mWall!!.setText(routeEntry.wall)
        mWall!!.visibility = View.VISIBLE
        mWallTV!!.visibility = View.GONE
        mNotes!!.setText(routeEntry.note)
        mNotes!!.visibility = View.VISIBLE
        mNotesTV!!.visibility = View.GONE
        editMenuCheck = false
        readyUpdateCheck = true
        editMenuItem!!.setTitle(R.string.update)
        mPhotoIntentButton!!.visibility = View.VISIBLE
    }

    fun refactorUIonUpdateRoute() {
        Toast.makeText(baseContext, "Your note has been updated", Toast.LENGTH_LONG).show()
        populateStaticUI(route)
        editMenuItem!!.title = "Edit"
        editMenuCheck = true
        readyUpdateCheck = false
    }

    fun getRadioBox(color: String?) {
        val newColourStr = color!!.toLowerCase()
        when (newColourStr) {
            "blue" -> (findViewById<View>(R.id.route_colour_group) as RadioGroup).check(R.id.blueButton)
            "pink" -> (findViewById<View>(R.id.route_colour_group) as RadioGroup).check(R.id.pinkButton)
            "orange" -> (findViewById<View>(R.id.route_colour_group) as RadioGroup).check(R.id.orangeButton)
            "yellow" -> (findViewById<View>(R.id.route_colour_group) as RadioGroup).check(R.id.yellowButton)
        }
    }

    private fun dispatchTakePictureIntent() {
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        // Ensure that there's a camera activity to handle the intent
        if (takePictureIntent.resolveActivity(packageManager) != null) {
            // Create the File where the photo should go
            var photoFile: File? = null
            photoFile = try {
                createImageFile()
            } catch (ex: IOException) {
                return
            }
            if (photoFile != null) {
                val photoURI = FileProvider.getUriForFile(Objects.requireNonNull(applicationContext),
                        BuildConfig.APPLICATION_ID + ".fileprovider", photoFile)
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
                startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
                climbWallUri = photoURI
            }
        }
    }

    private fun setSavedImageIfPresent() {
        if (route!!.getmImageLocation() != null) {
            climbWallUri = Uri.parse(route!!.getmImageLocation())
        }
    }

    private fun setImage() {
        if (climbWallUri != null) {
            Picasso.get().load(climbWallUri).into(wallPhotoImageView)
        } else {
            val url = "https://p2.piqsels.com/preview/385/249/229/rock-climber-rope-mounatin.jpg"
            Picasso.get().load(url).into(wallPhotoImageView)
        }
    }

    private fun delaySetImage() {
        try {
            TimeUnit.SECONDS.sleep(1)
            setImage()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == Activity.RESULT_OK) {
            setImage()
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    @Throws(IOException::class)
    private fun createImageFile(): File {
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
        val imageFileName = "JPEG_" + timeStamp + "_"
        val storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        val image = File.createTempFile(
                imageFileName,  /* prefix */
                ".jpg",  /* suffix */
                storageDir /* directory */
        )
        currentPhotoPath = image.absolutePath
        return image
    }

    companion object {
        const val EXTRA_ROUTE_ID = "extraTaskId"
        const val INSTANCE_ROUTE_ID = "instanceTaskId"
        private const val DEFAULT_ROUTE_ID = -1
        private const val DATE_FORMAT = "dd/MM/yyy"
        const val REQUEST_IMAGE_CAPTURE = 1
    }
}